### PR TITLE
reverseproxy: Add upstreams healthy metrics

### PR DESCRIPTION
**Pull request by david-szabo97**
Adds a new metric that tracks the health status of reverse proxy upstreams.

```
# HELP caddy_reverse_proxy_upstreams_healthy Health status of reverse proxy upstreams.
# TYPE caddy_reverse_proxy_upstreams_healthy gauge
caddy_reverse_proxy_upstreams_healthy{upstream="localhost:8080"} 1
caddy_reverse_proxy_upstreams_healthy{upstream="localhost:8081"} 0
```

**Metric**:
- Name: `caddy_reverse_proxy_upstreams_healthy`
- Type: `gauge`
- Values:
   - `1` - the upstream is healthy
   - `0` - the upstream is unhealthy
- Labels:
   - `upstream` - Address of the upstream

**Example Grafana panel**:
image

Partially resolves https://github.com/caddyserver/caddy/issues/4140

### CODE REVIEW

**Comment by mholt** [comment]
Looks cool! I'm not familiar with Prometheus so I'll let our in-house expert take a look if he has a chance :smiley:

I'll review the usage of the reverse proxy internals though.

**Comment by mervyd** [comment]
#4935

**Comment by mervyd** [comment]
#3014

File: modules/caddyhttp/reverseproxy/metrics.go
Comment (mholt) [inline modules/caddyhttp/reverseproxy/metrics.go:21]:
```suggestion
	const ns, sub = "caddy", "reverse_proxy"
```

Technically, the module name is `http.handlers.reverse_proxy` -- not sure how much we care to have namespace parity here?

**Review by mholt — COMMENTED** [review]
Looks pretty good, thanks for the change! I'll leave it to Dave to do the final review if he has a chance.

File: modules/caddyhttp/reverseproxy/metrics.go
Comment (david-szabo97) [reply modules/caddyhttp/reverseproxy/metrics.go:44]: We do have one in the goroutine. Since the `Init` function only contains a single goroutine call, I don't see the point to add a recover() in the `Init` function itself. I might misunderstand something 😄

File: modules/caddyhttp/reverseproxy/metrics.go
Comment (david-szabo97) [reply modules/caddyhttp/reverseproxy/metrics.go:21]: This makes sense. Internally it's `reverseproxy`, but externally it's always `reverse_proxy`. I'll change this to `reverse_proxy` to avoid confusing users. Better to stay with consistent naming.

File: modules/caddyhttp/reverseproxy/metrics.go
Comment (mholt) [reply modules/caddyhttp/reverseproxy/metrics.go:45]: Oh; wow, not sure how I missed that.

**Review by mholt — APPROVED** [review]

File: modules/caddyhttp/reverseproxy/metrics.go
Comment (david-szabo97) [reply modules/caddyhttp/reverseproxy/metrics.go:57]: I've tried to figure out how to make this configurable, but I'm kind of lost at this point.
Can you point me in the right direction?

I guess this should be in the server configuration block, or?

Comment (mholt) [reply modules/caddyhttp/reverseproxy/metrics.go:57]: That's a good question; I'd probably need to think about it a bit. Generally, the nearest/most specific place in the config as possible would be preferred.

It doesn't have to be a blocker for this PR, but we should choose a conservative value until it's configurable I think. Is 5-10s tolerable?

Comment (david-szabo97) [reply modules/caddyhttp/reverseproxy/metrics.go]: Sounds like a plan! Adjusted the interval to 10 seconds.

**Comment by mholt** [comment]
@hairyhenderson I know you're busy so I won't require a review from you to merge this! Let me know if you want me to hold off so you can review, otherwise I will probably just merge it in.

**Review by hairyhenderson — APPROVED** [review]
Thanks @david-szabo97, LGTM!

**Comment by mholt** [comment]
Great, thank you for the contribution @david-szabo97 !

**Comment by HeCorr** [comment]
I'm not a big fan of the hard-coded 10s delay between metric updates.
I would personally like a shorter delay so metrics (and consequently alarms) are more precise, thus a way to configure that would be greatly appreciated!
Awesome work nonetheless  :)

**Pull request merged**
